### PR TITLE
License component code

### PR DESCRIPTION
**Pull request by johnsimons**
Part of:

- https://github.com/Particular/ServiceControl/issues/4251

### CODE REVIEW

File: src/Particular.LicensingComponent/ThroughputCollector.cs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    public async Task<ReportGenerationState> GetReportGenerationState(CancellationToken cancellationToken)
+    {
+        var reportGenerationState = new ReportGenerationState(transport)
+        {
+            ReportCanBeGenerated = await dataStore.IsThereThroughputForLastXDays(30, cancellationToken)
+        };
+
+        if (!reportGenerationState.ReportCanBeGenerated)
+        {
+            reportGenerationState.Reason = "24hrs worth of data needs to exist in the last 30 days.";
```
Comment (johnsimons) [inline src/Particular.LicensingComponent/ThroughputCollector.cs:101]: I wonder if we can improve on this error?

**Review by johnsimons — COMMENTED** [review]
a few comments

File: src/Particular.LicensingComponent.UnitTests/Infrastructure/ThroughputCollectorTestFixture.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    using System.Threading.Tasks;
+    using Contracts;
+    using Microsoft.Extensions.DependencyInjection;
+    using NUnit.Framework;
+    using Persistence;
+
+    [TestFixture]
+    abstract class ThroughputCollectorTestFixture
+    {
+        //public Action<PersistenceSettings> SetPersistenceSettings = _ => { };
```
[the remark this answers is not in the document]
Comment (jpalac) [reply src/Particular.LicensingComponent.UnitTests/Infrastructure/ThroughputCollectorTestFixture.cs:13]: done

File: src/Particular.LicensingComponent.UnitTests/ServiceControl.runsettings [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<RunSettings>
+  <RunConfiguration>
+    <!-- The tests need to be run in x64 otherwise an exception about RavenDB (Voron) not being supported in 32bit mode will be thrown. -->
```
[the remark this answers is not in the document]
Comment (jpalac) [reply src/Particular.LicensingComponent.UnitTests/ServiceControl.runsettings:4]: removed - let's see if tests pass

File: src/Particular.LicensingComponent/ThroughputCollector.cs (hunk 1 above)
Comment (jpalac) [reply src/Particular.LicensingComponent/ThroughputCollector.cs:101]: Any suggestions @johnsimons ?

Comment (johnsimons) [inline src/Particular.LicensingComponent/ThroughputCollector.cs:101]:
```suggestion
            reportGenerationState.Reason = "24 hours worth of data needs to exist in the last 30 days.";
```

Comment (johnsimons) [reply src/Particular.LicensingComponent/ThroughputCollector.cs:101]: For context, this error message is displayed in ServicePulse
image

File: src/Particular.LicensingComponent/Shared/DictionaryExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿namespace Particular.LicensingComponent.Shared
+{
+    static class DictionaryExtensions
+    {
+        public static Dictionary<string, string> AddOrUpdate(this Dictionary<string, string> dictionary, string key, string value)
```
[the remark this answers is not in the document]
Comment (johnsimons) [reply src/Particular.LicensingComponent/Shared/DictionaryExtensions.cs]: Fixed

File: src/Particular.LicensingComponent/ThroughputCollector.cs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    public async Task<ReportGenerationState> GetReportGenerationState(CancellationToken cancellationToken)
+    {
+        var reportGenerationState = new ReportGenerationState(transport)
+        {
+            ReportCanBeGenerated = await dataStore.IsThereThroughputForLastXDays(30, cancellationToken)
+        };
+
+        if (!reportGenerationState.ReportCanBeGenerated)
+        {
+            reportGenerationState.Reason = "24 hours worth of data needs to exist in the last 30 days.";
```
Comment (abparticular) [inline src/Particular.LicensingComponent/ThroughputCollector.cs:101]:
```suggestion
            reportGenerationState.Reason = "There needs to be at least one 24 hour period of data collected in the last 30 days.";
```

**Review by abparticular — APPROVED** [review]

**Review by jpalac — APPROVED** [review]

**Pull request merged**